packages/ovm-toolchain/src/watcher.ts

```diff
@@ -0,0 +1,68 @@
+/* External Imports */
+import { ethers } from 'ethers-v4'
+
+interface Layer {
+  provider: any
+  messengerAddress: string
+}
+
+interface WatcherOptions {
+  l1: Layer
+  l2: Layer
+}
+
+export class Watcher {
+  public l1: Layer
+  public l2: Layer
+
+  constructor(opts: WatcherOptions) {
+    this.l1 = opts.l1
+    this.l2 = opts.l2
+  }
+
+  public async getMessageHashesFromL1Tx(l1TxHash: string): Promise<string[]> {
+    return this._getMessageHashesFromTx(true, l1TxHash)
+  }
+  public async getMessageHashesFromL2Tx(l2TxHash: string): Promise<string[]> {
+    return this._getMessageHashesFromTx(false, l2TxHash)
+  }
+
+  public onceL2Relay(msgHash: string, callback: Function) {
+    return this._onceRelay(false, msgHash, callback)
+  }
+
+  public onceL1Relay(msgHash: string, callback: Function) {
```

Comment:
What were the thoughts here about using a callback vs just returning a promise?

Reply:
Oh also should probably add a return value for this function

Reply:
Was trying to emulate the `provider.once` functionality in `ethers` https://docs.ethers.io/v5/single-page/#/v5/api/providers/provider/-%23-Provider-once -

Reply:
Another way to do this that is a bit more flexible I think is to make the `Watcher` an `EventEmitter` and then update this part of the code:

```js
    layer.provider.on(filter, (log: any) => {
      if (log.data === msgHash) {
        this.emit('relayed', log.transactionHash)
      }
```

Then the `Watcher` would have the `on` property and the `once` property as well as the rest of the highly used/tested `EventEmitter` properties